File: src/audit.ts
import { appendFileSync, mkdirSync, existsSync } from "node:fs";
import { resolve } from "node:path";

const AUDIT_DIR = resolve(process.cwd(), "data", "audit");

interface AuditEntry {
  taskId: string;
  action: string;
  who: string;
  channel: string;
  detail?: Record<string, unknown>;
}

let initialized = false;

function ensureDir(): void {
  if (initialized) return;
  if (!existsSync(AUDIT_DIR)) {
    mkdirSync(AUDIT_DIR, { recursive: true });
  }
  initialized = true;
}

/**
 * Append a structured audit log entry in JSONL format.
 * Each line: { ts, taskId, action, who, channel, detail }
 *
 * Audit files are rotated daily: audit-YYYY-MM-DD.jsonl
 */
export function auditLog(entry: AuditEntry): void {
  ensureDir();

  const now = new Date();
  const dateStr = now.toISOString().slice(0, 10);
  const filePath = resolve(AUDIT_DIR, `audit-${dateStr}.jsonl`);

  const line = JSON.stringify({
    ts: now.toISOString(),
    taskId: entry.taskId,
    action: entry.action,
    who: entry.who,
    channel: entry.channel,
    ...(entry.detail ? { detail: entry.detail } : {}),
  });

  try {
    appendFileSync(filePath, line + "\n");
  } catch {
    // Fail silently - audit should never crash the main process
  }
}

File: src/utils/audit.py
"""JSONL audit logger, matching TS audit.ts. Daily-rotated files."""

from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path

AUDIT_DIR = Path.cwd() / "data" / "audit"

_initialized = False


def _ensure_dir() -> None:
    global _initialized
    if _initialized:
        return
    AUDIT_DIR.mkdir(parents=True, exist_ok=True)
    _initialized = True


def audit_log(
    *,
    task_id: str,
    action: str,
    who: str,
    channel: str,
    detail: dict | None = None,
) -> None:
    _ensure_dir()

    now = datetime.now(timezone.utc)
    date_str = now.strftime("%Y-%m-%d")
    file_path = AUDIT_DIR / f"audit-{date_str}.jsonl"

    entry: dict = {
        "ts": now.isoformat(),
        "taskId": task_id,
        "action": action,
        "who": who,
        "channel": channel,
    }
    if detail:
        entry["detail"] = detail

    try:
        with open(file_path, "a", encoding="utf-8") as f:
            f.write(json.dumps(entry, ensure_ascii=False) + "\n")
    except Exception:
        pass
